File: src/main/kotlin/mapServer/MapServer.java
package mapServer;

import com.sun.net.httpserver.HttpContext;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.stream.Stream;

public class MapServer extends HttpServer {
    public MapServer(){
        createContext("/getClosestNode", exchange -> {
            InputStream input = exchange.getRequestBody();
            OutputStream output = exchange.getResponseBody();

            Map<String,String> parms = queryToMap(exchange.getRequestURI().getQuery());
            double lon = Double.parseDouble(parms.get("lon"));
            double lat = Double.parseDouble(parms.get("lat"));

            output.write(("hi lon: " + parms.get("lon") + "lat: " +  parms.get("lat")).getBytes(StandardCharsets.UTF_8));
            output.close();

            create(new InetSocketAddress(8080), 0);
            start();
        });
        
    }
    public static Map<String, String> queryToMap(String query){
        Map<String, String> result = new HashMap<String, String>();
        for (String param : query.split("&")) {
            String pair[] = param.split("=");
            if (pair.length>1) {
                result.put(pair[0], pair[1]);
            }else{
                result.put(pair[0], "");
            }
        }
        return result;
    }
    @Override
    public void bind(InetSocketAddress addr, int backlog) throws IOException {

    }

    @Override
    public void start() {

    }

    @Override
    public void setExecutor(Executor executor) {

    }

    @Override
    public Executor getExecutor() {
        return null;
    }

    @Override
    public void stop(int delay) {

    }

    @Override
    public HttpContext createContext(String path, HttpHandler handler) {
        return null;
    }

    @Override
    public HttpContext createContext(String path) {
        return null;
    }

    @Override
    public void removeContext(String path) throws IllegalArgumentException {

    }

    @Override
    public void removeContext(HttpContext context) {

    }

    @Override
    public InetSocketAddress getAddress() {
        return null;
    }

}

File: src/main/kotlin/mapServer/MapServer.kt
package mapServer

import closestNode.getClosestNodeNaively
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import graph.Graph
import pathfind.Dijkstra
import java.io.File
import java.lang.IllegalArgumentException
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import kotlin.collections.HashMap

class MapServer {

    companion object {

        private const val path = "html\\html test.html"
        private const val dancingBananaPath = "html\\banana_anim_custom.png"
        private const val thinkingBananaPath = "html\\thinking_banana.png"

        private val server = HttpServer.create(InetSocketAddress(8080), 0)

        fun setupServer()
        {
            println("setting up server at address ${server.address}")
            contextMain()
            contextClosestNode()
            contextDancingBanana()
            contextThinkingBanana()
            server.start()
            println("server started!")
        }

        private fun contextMain() {
            println("creating main context...")
            server.createContext("/") { exchange: HttpExchange ->
                exchange.sendResponseHeaders(200, File(path).length())
                val output = exchange.responseBody
                output.write(File(path).readBytes())
                output.close()
            }
        }

        private fun contextClosestNode() {
            println("creating closest node context...")
            server.createContext("/getClosestNode") { exchange: HttpExchange ->
                val parms = queryToMap(exchange.requestURI.query)
                val closestNode = getClosestNodeNaively(parms["lat"]!!.toDouble(), parms["lon"]!!.toDouble())
                var response = ""+Graph.lat[closestNode]+","+Graph.lon[closestNode]+","
                val output = exchange.responseBody

                // automatically starts the Dijkstra algorithm at the input location
                if (parms["amountMarkers"]!!.toInt() == 0) {
                    println("starting Dijkstra...")
                    Dijkstra(closestNode)
                    println("finished Dijkstra")
                    exchange.sendResponseHeaders(200, response.toByteArray(StandardCharsets.UTF_8).size.toLong())
                    output.write(response.toByteArray(StandardCharsets.UTF_8))
                }
                // returns the path to the target location
                else if (parms["amountMarkers"]!!.toInt() == 1) {

                    val currentPath = Graph.returnPath(closestNode)

                    while (!currentPath.isEmpty()) {
                        response += Graph.lat[currentPath[0]].toString() + ","
                        response += Graph.lon[currentPath[0]].toString() + ","
                        currentPath.removeFirst()
                    }
                    exchange.sendResponseHeaders(200, response.toByteArray(StandardCharsets.UTF_8).size.toLong())
                    output.write(response.toByteArray(StandardCharsets.UTF_8))
                }
                else throw IllegalArgumentException("amountMarkers have to be 0 or 1!")

                output.close()
            }
        }

        private fun contextDancingBanana() {
            println("creating dancing banana context...")
            server.createContext("/dancingBanana") { exchange: HttpExchange ->
                exchange.sendResponseHeaders(200, File(dancingBananaPath).length())
                val output = exchange.responseBody
                //println("BANANA IS MF DANCING!")
                output.write(File(dancingBananaPath).readBytes())
                output.close()
            }
        }

        private fun contextThinkingBanana() {
            println("creating thinking banana context...")
            server.createContext("/thinkingBanana") { exchange: HttpExchange ->
                exchange.sendResponseHeaders(200, File(thinkingBananaPath).length())
                val output = exchange.responseBody
                //println("Da banana must do the thinking...")
                output.write(File(thinkingBananaPath).readBytes())
                output.close()
            }
        }


        private fun queryToMap(query: String): Map<String, String> {
            val result: MutableMap<String, String> = HashMap()
            for (param in query.split("&").toTypedArray()) {
                val pair = param.split("=").toTypedArray()
                if (pair.size > 1) {
                    result[pair[0]] = pair[1]
                } else {
                    result[pair[0]] = ""
                }
            }
            return result
        }


    }
}